File: src/cs/IdpProtocol/IncomingTransaction.cs
// Copyright (c) VitalElement. All rights reserved.
// Licensed under the MIT license. See licence.md file in the project root for full license information.

using MiscUtil.Conversion;
using System;
using System.IO;
using System.Runtime.InteropServices;
using System.Text;

namespace IdpProtocol
{
    public class IncomingTransaction
    {
        public int _readIndex;
        private readonly IdpPacket _packet;

        internal IncomingTransaction(IdpPacket packet)
        {
            _packet = packet;

            Data = packet.Data;
            _readIndex = 10;

            CommandId = Read<UInt16>();
            TransactionId = Read<UInt32>();
            Flags = Read<IdpCommandFlags>();

            Source = packet.Source;
            Destination = packet.Destination;
        }

        public UInt16 Source { get; }

        public UInt16 Destination { get; }

        public byte[] Data { get; }

        /*public Span<byte> SpanToEnd()
        {
            return new Span<byte>(_data, _readIndex, _data.Length - _readIndex);
        }*/

        public byte[] BufferToEnd ()
        {
            var result = new byte[Data.Length - _readIndex];

            Array.Copy(Data, _readIndex, result, 0, result.Length);

            return result;
        }

        public IdpCommandFlags Flags { get; }
        public UInt16 CommandId { get; }
        public UInt32 TransactionId { get; }

        public Stream StreamToEnd()
        {
            var result = new MemoryStream(Data, _readIndex, Data.Length - _readIndex);

            return result;
        }

        public string ReadUtf8String()
        {
            int index = _readIndex;
            bool stringFound = false;

            while (index < Data.Length)
            {
                if (Data[index] == '\0')
                {
                    stringFound = true;
                    break;
                }

                index++;
            }

            if (stringFound)
            {
                var result = Encoding.UTF8.GetString(Data, _readIndex, index - _readIndex);

                _readIndex += index - _readIndex;

                return result;
            }
            else
            {
                return "";
            }
        }

        public T Read<T>() where T : struct
        {
            object result = null;
            bool increment = true;

            if (typeof(T) == typeof(bool))
            {
                result = EndianBitConverter.Big.ToBoolean(Data, _readIndex);

                increment = false;

                _readIndex++;
            }
            else if (typeof(T) == typeof(Int32))
            {
                result = EndianBitConverter.Big.ToInt32(Data, _readIndex);
            }
            else if (typeof(T) == typeof(UInt32))
            {
                result = EndianBitConverter.Big.ToUInt32(Data, _readIndex);
            }
            else if (typeof(T) == typeof(Int16))
            {
                result = EndianBitConverter.Big.ToInt16(Data, _readIndex);
            }
            else if (typeof(T) == typeof(UInt16))
            {
                result = EndianBitConverter.Big.ToUInt16(Data, _readIndex);
            }
            else if (typeof(T) == typeof(char))
            {
                result = EndianBitConverter.Big.ToChar(Data, _readIndex);
            }
            else if (typeof(T) == typeof(byte))
            {
                result = Data[_readIndex];
            }
            else if (typeof(T) == typeof(float))
            {
                result = EndianBitConverter.Big.ToSingle(Data, _readIndex);
            }
            else if (typeof(T) == typeof(IdpCommandFlags))
            {
                result = (IdpCommandFlags)Data[_readIndex];

                increment = false;

                _readIndex++;
            }
            else if (typeof(T) == typeof(IdpResponseCode))
            {
                result = (IdpResponseCode)Data[_readIndex];

                increment = false;

                _readIndex++;
            }
            else if (typeof(T) == typeof(Guid))
            {
                result = new Guid(Read<int>(), Read<short>(), Read<byte>(),
                    Read<byte>(), Read<byte>(), Read<byte>(), 
                    Read<byte>(), Read<byte>(), Read<byte>(), 
                    Read<byte>(), Read<byte>());

                increment = false;
            }
            else
            {
                throw new NotSupportedException($"Unable to decode type {typeof(T).FullName} from response.");
            }

            if (increment)
            {
                _readIndex += Marshal.SizeOf(result);
            }

            return (T)result;
        }
    }
}

File: src/cpp/IncomingTransaction.cpp
// Copyright (c) VitalElement. All rights reserved.
// Licensed under the MIT license. See licence.md file in the project root for
// full license information.
#include "IncomingTransaction.h"
#include <cstring>

IncomingTransaction::IncomingTransaction (std::shared_ptr<IdpPacket> packet)
{
    _packet = packet;

    _data = packet->Buffer ();
    _readIndex = 10;
    _readLimit = packet->Length ();

    if (((uint8_t) packet->Flags () & (uint8_t) IdpFlags::CRC) ==
        (uint8_t) IdpFlags::CRC)
    {
        if (_readLimit < 4)
        {
            ThrowException (-1, "Invalid IDP packet length");
        }

        _readLimit -= 4;
    }

    if (_readLimit < 11)
    {
        ThrowException (-1, "Invalid IDP packet length");
    }

    // Exclude ETX from transaction payload reads.
    _readLimit -= 1;

    _commandId = Read<uint16_t> ();
    _transactionId = Read<uint32_t> ();
    _flags = (IdpCommandFlags) Read<uint8_t> ();

    _source = packet->Source ();
    _destination = packet->Destination ();
}

IncomingTransaction::~IncomingTransaction ()
{
}

std::shared_ptr<IdpPacket> IncomingTransaction::Packet ()
{
    return _packet;
}

IdpCommandFlags IncomingTransaction::Flags ()
{
    return _flags;
}

uint16_t IncomingTransaction::CommandId ()
{
    return _commandId;
}

uint32_t IncomingTransaction::TransactionId ()
{
    return _transactionId;
}

uint16_t IncomingTransaction::Source ()
{
    return _source;
}

uint16_t IncomingTransaction::Destination ()
{
    return _destination;
}

const char* IncomingTransaction::ReadCString ()
{
    if (_readIndex > _readLimit)
    {
        ThrowException (-1, "IDP packet read out of bounds");
    }

    auto remaining = _readLimit - _readIndex;
    auto terminator = static_cast<const char*> (
        memchr (_data.get () + _readIndex, '\0', remaining));

    if (terminator == nullptr)
    {
        ThrowException (-1, "Unterminated string in IDP packet");
    }

    auto length = (uint32_t) (terminator - (const char*) (_data.get () + _readIndex)) + 1;

    auto result = new char[length];

    Read (result, length);

    return result;
}

Guid_t IncomingTransaction::ReadGuid ()
{
    auto data1 = Read<uint32_t> ();
    auto data2 = Read<uint16_t> ();
    auto data3 = Read<uint16_t> ();
    uint8_t data4[8];
    Read (data4, 8);

    return Guid_t (data1, data2, data3, data4);
}

void IncomingTransaction::Read (void* destination, uint32_t length)
{
    EnsureReadable (length);
    memcpy (destination, (void*) (_data.get () + _readIndex), length);
    _readIndex += length;
}

void* IncomingTransaction::ConsumeData(uint32_t length)
{
    EnsureReadable (length);
    void* data = (void *) (_data.get() + _readIndex);
    _readIndex += length;

    return data;
}

void IncomingTransaction::EnsureReadable (uint32_t length)
{
    if (_readIndex > _readLimit || length > (_readLimit - _readIndex))
    {
        ThrowException (-1, "IDP packet read out of bounds");
    }
}
